docs/design/specs/Ecma-335-Augments.md

```diff
@@ -325,3 +325,19 @@ We propose the limit on #String and #Blob heap size is 2^29 (0.5 GB), that is an
 2) 0.5 GB is a very large heap. Having such a big PE file seems unreasonable and very rare scenario (if it exists at all).
 
 3) Having 3 spare bits available is very beneficial for the implementation. It allows to represent WinRT projected strings, namespaces, etc. in very efficient way. If we had to represent heap indices with all 32 bits it would bloat various structures and increase memory pressure. PE files over 0.5 GB of size are very rare, but the overhead would affect all compilers and tools working with the metadata reader.
+
+## Module Initializer
+
+All modules may have a module initializer. This method shall be static, a member of the module, take no parameters, return no value, be marked with **rtspecialname** and **specialname**, and be named `.cctor`.
+
+There are no limitations on what code is permitted in a module initializer. Module initializers are permitted to run and call both managed and unmanaged code.
+
+### Module Initialization Guarantees
+
+The CLI shall provide the following guarantees regarding module initialization:
+
+1. A module initializer is executed at, or sometime before, first access to any types, methods, or data defined in the module.
+
+2. A module initializer shall run exactly once for any given module unless explicitly called by user code.
+
+3. No method other than those called directly or indirectly from the module initializer will be able to access the types, methods, or data in a module before its initializer completes execution.
```

Comment:
1 and 3 *sound like* they would include access via reflection. However, `Assembly.GetTypes` currently returns type instances without triggering the module initializer of the module that the types came from. This is troublesome for plugin situations that need to hook AssemblyResolve/AssemblyLoadContext.Resolving before any of the types get a chance to load, as in the case of VSTest adapters.

1. Can the runtime be fixed?

2. Should reflection be explicitly mentioned here?

Reply:
This is using similar language regular constructors (e.g. look for "first access to any static field").

Acquiring reflection object does not count as trigger for regular constructors, so I do not think it should count as trigger for module constructors either. Fixing it would be a potential breaking change - the code out there may be depending on the fact that loading an assembly via Assembly.Load does not actually run any code in it.

Reply:
Thanks. Would you consider reflection to be something worth mentioning at all, or is it too obvious if you know the context?

Reply:
I do not think this needs to mention reflection.